File: Java/Trees/Van Emde Boas.java
import java.util.Arrays;

public class MyVEBTree {
    private int universeSize;
    private int minimum;
    private int maximum;
    private MyVEBTree summary;
    private MyVEBTree[] cluster;

    public MyVEBTree(int size) {
        universeSize = size;
        minimum = -1;
        maximum = -1;

        if (size > 2) {
            int upperSize = (int) Math.ceil(Math.sqrt(size));
            int lowerSize = (int) Math.floor(Math.sqrt(size));

            summary = new MyVEBTree(upperSize);
            cluster = new MyVEBTree[upperSize];

            for (int i = 0; i < upperSize; i++)
                cluster[i] = new MyVEBTree(lowerSize);
        }
    }

    public void insert(int key) {
        if (minimum == -1) {
            minimum = maximum = key;
            return;
        }

        if (key < minimum) {
            int temp = key;
            key = minimum;
            minimum = temp;
        }

        if (universeSize > 2) {
            if (cluster[key / (int) Math.sqrt(universeSize)].minimum == -1)
                summary.insert(key / (int) Math.sqrt(universeSize));

            cluster[key / (int) Math.sqrt(universeSize)].insert(key % (int) Math.sqrt(universeSize));
        }

        if (key > maximum)
            maximum = key;
    }

    public void remove(int key) {
        if (minimum == maximum) {
            minimum = maximum = -1;
            return;
        }

        if (universeSize == 2) {
            if (key == 0)
                minimum = 1;
            else
                minimum = 0;

            maximum = minimum;
            return;
        }

        if (key == minimum) {
            int firstCluster = summary.minimum;
            key = firstCluster * (int) Math.sqrt(universeSize) + cluster[firstCluster].minimum;
            minimum = key;
        }

        cluster[key / (int) Math.sqrt(universeSize)].remove(key % (int) Math.sqrt(universeSize));

        if (cluster[key / (int) Math.sqrt(universeSize)].minimum == -1)
            summary.remove(key / (int) Math.sqrt(universeSize));

        if (key == maximum) {
            if (summary.minimum == -1)
                maximum = minimum;
            else {
                int lastCluster = summary.maximum;
                maximum = lastCluster * (int) Math.sqrt(universeSize) + cluster[lastCluster].maximum;
            }
        }
    }

    public int successor(int key) {
        if (universeSize == 2) {
            if (key == 0 && maximum == 1)
                return 1;
            else
                return -1;
        }

        if (minimum != -1 && key < minimum)
            return minimum;

        int maxLow = cluster[key / (int) Math.sqrt(universeSize)].maximum;

        if (maxLow != -1 && key % (int) Math.sqrt(universeSize) < maxLow) {
            int offset = cluster[key / (int) Math.sqrt(universeSize)].successor(key % (int) Math.sqrt(universeSize));
            return key / (int) Math.sqrt(universeSize) * (int) Math.sqrt(universeSize) + offset;
        }

        int succCluster = summary.successor(key / (int) Math.sqrt(universeSize));

        if (succCluster == -1)
            return -1;

        int offset = cluster[succCluster].minimum;
        return succCluster * (int) Math.sqrt(universeSize) + offset;
    }

    public int predecessor(int key) {
        if (universeSize == 2) {
            if (key == 1 && minimum == 0)
                return 0;
            else
                return -1;
        }

        if (maximum != -1 && key > maximum)
            return maximum;

        int minLow = cluster[key / (int) Math.sqrt(universeSize)].minimum;

        if (minLow != -1 && key % (int) Math.sqrt(universeSize) > minLow) {
            int offset = cluster[key / (int) Math.sqrt(universeSize)].predecessor(key % (int) Math.sqrt(universeSize));
            return key / (int) Math.sqrt(universeSize) * (int) Math.sqrt(universeSize) + offset;
        }

        int predCluster = summary.predecessor(key / (int) Math.sqrt(universeSize));

        if (predCluster == -1) {
            if (minimum != -1 && key > minimum)
                return minimum;
            else
                return -1;
        }

        int offset = cluster[predCluster].maximum;
        return predCluster * (int) Math.sqrt(universeSize) + offset;
    }

    public boolean contains(int key) {
        if (key == minimum || key == maximum)
            return true;

        if (universeSize == 2)
            return false;

        return cluster[key / (int) Math.sqrt(universeSize)].contains(key % (int) Math.sqrt(universeSize));
    }

    public static void main(String[] args) {
        MyVEBTree vebTree = new MyVEBTree(16);
        vebTree.insert(4);
        vebTree.insert(1);
        vebTree.insert(8);
        vebTree.insert(10);

        System.out.println(vebTree.contains(4)); // Output: true
        System.out.println(vebTree.contains(5)); // Output: false

        System.out.println(vebTree.successor(4)); // Output: 8
        System.out.println(vebTree.predecessor(4)); // Output: 1

        vebTree.remove(4);
        System.out.println(vebTree.contains(4)); // Output: false
    }
}

}

File: Python/Trees/Van Emde Boas.py
import math


class VEBTree:
    def __init__(self, size):
        self.universe_size = size
        self.minimum = -1
        self.maximum = -1
        self.summary = None
        self.cluster = []

        if size > 2:
            upper_size = math.ceil(math.sqrt(size))
            lower_size = math.floor(math.sqrt(size))

            self.summary = VEBTree(upper_size)
            self.cluster = [VEBTree(lower_size) for _ in range(upper_size)]

    def insert(self, key):
        if self.minimum == -1:
            self.minimum = self.maximum = key
            return

        if key < self.minimum:
            key, self.minimum = self.minimum, key

        if self.universe_size > 2:
            if self.cluster[key // int(math.sqrt(self.universe_size))].minimum == -1:
                self.summary.insert(key // int(math.sqrt(self.universe_size)))

            self.cluster[key // int(math.sqrt(self.universe_size))].insert(key % int(math.sqrt(self.universe_size)))

        if key > self.maximum:
            self.maximum = key

    def remove(self, key):
        if self.minimum == self.maximum:
            self.minimum = self.maximum = -1
            return

        if self.universe_size == 2:
            if key == 0:
                self.minimum = 1
            else:
                self.minimum = 0

            self.maximum = self.minimum
            return

        if key == self.minimum:
            first_cluster = self.summary.minimum
            key = first_cluster * int(math.sqrt(self.universe_size)) + self.cluster[first_cluster].minimum
            self.minimum = key

        self.cluster[key // int(math.sqrt(self.universe_size))].remove(key % int(math.sqrt(self.universe_size)))

        if self.cluster[key // int(math.sqrt(self.universe_size))].minimum == -1:
            self.summary.remove(key // int(math.sqrt(self.universe_size)))

        if key == self.maximum:
            if self.summary.minimum == -1:
                self.maximum = self.minimum
            else:
                last_cluster = self.summary.maximum
                self.maximum = last_cluster * int(math.sqrt(self.universe_size)) + self.cluster[last_cluster].maximum

    def successor(self, key):
        if self.universe_size == 2:
            if key == 0 and self.maximum == 1:
                return 1
            else:
                return -1

        if self.minimum != -1 and key < self.minimum:
            return self.minimum

        max_low = self.cluster[key // int(math.sqrt(self.universe_size))].maximum

        if max_low != -1 and key % int(math.sqrt(self.universe_size)) < max_low:
            offset = self.cluster[key // int(math.sqrt(self.universe_size))].successor(
                key % int(math.sqrt(self.universe_size)))
            return key // int(math.sqrt(self.universe_size)) * int(math.sqrt(self.universe_size)) + offset

        succ_cluster = self.summary.successor(key // int(math.sqrt(self.universe_size)))

        if succ_cluster == -1:
            return -1

        offset = self.cluster[succ_cluster].minimum
        return succ_cluster * int(math.sqrt(self.universe_size)) + offset

    def predecessor(self, key):
        if self.universe_size == 2:
            if key == 1 and self.minimum == 0:
                return 0
            else:
                return -1

        if self.maximum != -1 and key > self.maximum:
            return self.maximum

        min_low = self.cluster[key // int(math.sqrt(self.universe_size))].minimum

        if min_low != -1 and key % int(math.sqrt(self.universe_size)) > min_low:
            offset = self.cluster[key // int(math.sqrt(self.universe_size))].predecessor(
                key % int(math.sqrt(self.universe_size)))
            return key // int(math.sqrt(self.universe_size)) * int(math.sqrt(self.universe_size)) + offset

        pred_cluster = self.summary.predecessor(key // int(math.sqrt(self.universe_size)))

        if pred_cluster == -1:
            if self.minimum != -1 and key > self.minimum:
                return self.minimum
            else:
                return -1

        offset = self.cluster[pred_cluster].maximum
        return pred_cluster * int(math.sqrt(self.universe_size)) + offset

    def contains(self, key):
        if key == self.minimum or key == self.maximum:
            return True

        if self.universe_size == 2:
            return False

        return self.cluster[key // int(math.sqrt(self.universe_size))].contains(
            key % int(math.sqrt(self.universe_size)))


veb_tree = VEBTree(16)
veb_tree.insert(4)
veb_tree.insert(1)
veb_tree.insert(8)
veb_tree.insert(10)

print(veb_tree.contains(4))  # Output: True
print(veb_tree.contains(5))  # Output: False

print(veb_tree.successor(4))  # Output: 8
print(veb_tree.predecessor(4))  # Output: 1

veb_tree.remove(4)
print(veb_tree.contains(4))  # Output: False
